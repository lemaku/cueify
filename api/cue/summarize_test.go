package cue

import (
	"cuelang.org/go/cue/cuecontext"
	"encoding/json"
	"github.com/gkampitakis/go-snaps/snaps"
	"testing"
)

func matchSummarizeSnapshot(t *testing.T, value interface{}) {
	jsonResult, err := json.MarshalIndent(value, "", "    ")
	if err != nil {
		panic("Error: couldn't serialize result to use for the snapshot")
	}
	snaps.MatchSnapshot(t, string(jsonResult))
}

func TestSummarizeEmpty(t *testing.T) {
	const schema = `
#student: {
	matNr:  string & =~"^[0-9]{8}$"
	name:   string
	active: *true | bool
    if active {
        semester: int
    }
}

#universities: {
	tuwien: {
		name: "Vienna University of Technology",
		students: [...#student]
	},
	countryCode: string
}

#export: #universities
`
	matchSummarizeSnapshot(t, Summarize("{}", schema))
}

func TestSummarizeTopLevelArray(t *testing.T) {
	const schema = `
#student: {
	matNr:  string & =~"^[0-9]{8}$"
	name:   string
	active: *true | bool
    if active {
        semester: int
    }
}

#export: [...#student]
`
	matchSummarizeSnapshot(t, Summarize("[]", schema))
	matchSummarizeSnapshot(t, Summarize("[{}]", schema))
	matchSummarizeSnapshot(t, Summarize(`[{ name: "Max Mustermann"}]`, schema))
	matchSummarizeSnapshot(t, Summarize(`[{ name: "Max Mustermann", matNr: "11111111"}]`, schema))
	matchSummarizeSnapshot(t, Summarize(`[{ name: "Max Mustermann", matNr: "11111111", semester: 5}]`, schema))
}

func TestDependencyHeuristic(t *testing.T) {
	raw := `
{
	z: y
	y: x
	x: int
} & {}`
	context := cuecontext.New()
	value := context.CompileString(raw)

	depMap, _ := analyzeDeps(value, []string{"x", "y", "z"})
	matchSummarizeSnapshot(t, depMap)
}

func TestDependencyHeuristicWithConcreteValues(t *testing.T) {
	raw := `
{
	z: y
	y: x
	x: t
	t: int
} & { x: 4 }`
	context := cuecontext.New()
	value := context.CompileString(raw)

	depMap, _ := analyzeDeps(value, []string{"t", "x", "y", "z"})
	matchSummarizeSnapshot(t, depMap)
}

func TestDependencyHeuristicWithCycles(t *testing.T) {
	raw := `
{
	x: y
	y: x
} & {}`
	context := cuecontext.New()
	value := context.CompileString(raw)

	_, err := analyzeDeps(value, []string{"x", "y", "z"})
	snaps.MatchSnapshot(t, err.Error())
}

func TestPartialExportShouldNotIncludeOptionalProperties(t *testing.T) {
	// "a" should not be included since it is optional
	matchSummarizeSnapshot(t, Summarize(`{ c: 1.3 }`, `
#export: {
	a?: null,
	b: string,
	c: float
}`).Value)
	// now "a" is not optional anymore and has concrete value, so should be included
	matchSummarizeSnapshot(t, Summarize(`{ c: 1.3 }`, `
#export: {
	a: null,
	b: string,
	c: float
}`).Value)
}

func TestPartialExportShouldIncludeDefaults(t *testing.T) {
	matchSummarizeSnapshot(t, Summarize(`{}`, `
#export: {
	a: string | *"this is the default",
}`).Value)
}

func TestPartialExportShouldHandleStructsAndListsCorrectly(t *testing.T) {
	const schema = `
#export: {
	a: { b: string },
	c?: { d: string },
	e: [...string],
	f?: [...string],
}`

	matchSummarizeSnapshot(t, Summarize(`{}`, schema).Value)
	matchSummarizeSnapshot(t, Summarize(`{c: {d: "hello"}, f: [",", "world"]}`, schema).Value)
}
