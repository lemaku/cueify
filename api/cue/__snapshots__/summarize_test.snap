
[TestSummarizeEmpty - 1]
{
    "value": {
        "tuwien": {
            "name": "Vienna University of Technology",
            "students": []
        }
    },
    "valid": false,
    "errors": [
        {
            "path": [
                "countryCode"
            ],
            "errors": [
                "incomplete value string"
            ]
        }
    ]
}
---

[TestSummarizeTopLevelArray - 1]
{
    "value": [],
    "valid": true,
    "errors": []
}
---

[TestSummarizeTopLevelArray - 2]
{
    "value": [
        {
            "active": true
        }
    ],
    "valid": false,
    "errors": [
        {
            "path": [
                "0",
                "matNr"
            ],
            "errors": [
                "incomplete value =~\"^[0-9]{8}$\""
            ]
        },
        {
            "path": [
                "0",
                "name"
            ],
            "errors": [
                "incomplete value string"
            ]
        },
        {
            "path": [
                "0",
                "semester"
            ],
            "errors": [
                "incomplete value int"
            ]
        }
    ]
}
---

[TestSummarizeTopLevelArray - 3]
{
    "value": [
        {
            "active": true,
            "name": "Max Mustermann"
        }
    ],
    "valid": false,
    "errors": [
        {
            "path": [
                "0",
                "matNr"
            ],
            "errors": [
                "incomplete value =~\"^[0-9]{8}$\""
            ]
        },
        {
            "path": [
                "0",
                "semester"
            ],
            "errors": [
                "incomplete value int"
            ]
        }
    ]
}
---

[TestSummarizeTopLevelArray - 4]
{
    "value": [
        {
            "active": true,
            "matNr": "11111111",
            "name": "Max Mustermann"
        }
    ],
    "valid": false,
    "errors": [
        {
            "path": [
                "0",
                "semester"
            ],
            "errors": [
                "incomplete value int"
            ]
        }
    ]
}
---

[TestSummarizeTopLevelArray - 5]
{
    "value": [
        {
            "active": true,
            "matNr": "11111111",
            "name": "Max Mustermann",
            "semester": 5
        }
    ],
    "valid": true,
    "errors": []
}
---

[TestPartialExport - 1]
{
    "value": {
        "a": null,
        "b": "abc",
        "c": 1.3,
        "d": 1,
        "g": []
    },
    "valid": false,
    "errors": [
        {
            "path": [
                "e"
            ],
            "errors": [
                "incomplete value bytes"
            ]
        }
    ]
}
---

[TestPartialExportShouldNotIncludeOptionalProperties - 1]
{
    "c": 1.3
}
---

[TestPartialExportShouldNotIncludeOptionalProperties - 2]
{
    "a": null,
    "c": 1.3
}
---

[TestPartialExportShouldIncludeDefaults - 1]
{
    "a": "this is the default"
}
---

[TestPartialExportShouldHandleStructsAndListsCorrectly - 1]
{
    "a": {},
    "e": []
}
---

[TestPartialExportShouldHandleStructsAndListsCorrectly - 2]
{
    "a": {},
    "c": {
        "d": "hello"
    },
    "e": [],
    "f": [
        ",",
        "world"
    ]
}
---

[TestPartialExportShouldHandleBytesCorrectly - 1]
{}
---

[TestDependencyHeuristic - 1]
{
    "x": 0,
    "y": 1,
    "z": 2
}
---

[TestDependencyHeuristicWithCycles - 1]
WasmAPI.Summarize: Dependency analysis failed - cycle detected
---

[TestDependencyHeuristicWithConcreteValues - 1]
{
    "t": 0,
    "x": 0,
    "y": 0,
    "z": 0
}
---
